trie-db/src/lib.rs

```diff
@@ -377,52 +403,69 @@ where
 /// Trait with definition of trie layout.
 /// Contains all associated trait needed for
 /// a trie definition or implementation.
-pub trait TrieLayout {
+/// Structure implementing this trait can define contextual behavior
+/// for trie, but should be small (clone cost should be insignifiant).
+pub trait TrieLayout: Default + Clone {
 	/// If true, the trie will use extension nodes and
 	/// no partial in branch, if false the trie will only
 	/// use branch and node with partials in both.
 	const USE_EXTENSION: bool;
 	/// If true, the trie will allow empty values into `TrieDBMut`
 	const ALLOW_EMPTY: bool = false;
+
 	/// Hasher to use for this trie.
 	type Hash: Hasher;
 	/// Codec to use (needs to match hasher and nibble ops).
 	type Codec: NodeCodec<HashOut=<Self::Hash as Hasher>::Out>;
+
+	/// Threshold above which an external node should be
+	/// use to store a node value.
+	fn max_inline_value(&self) -> Option<u32>;
+}
+
+/// Small enum indicating representation of a given children.
+pub enum ChildrenDecoded {
+	/// Children is stored as a node.
+	Hash,
+	/// Children is stored as an inline node.
+	Inline,
+	/// No children is present.
+	None,
 }
 
 /// This trait associates a trie definition with preferred methods.
 /// It also contains own default implementations and can be
 /// used to allow switching implementation.
 pub trait TrieConfiguration: Sized + TrieLayout {
 	/// Operation to build a trie db from its ordered iterator over its key/values.
-	fn trie_build<DB, I, A, B>(db: &mut DB, input: I) -> <Self::Hash as Hasher>::Out where
-	DB: HashDB<Self::Hash, usize>,
-	I: IntoIterator<Item = (A, B)>,
-	A: AsRef<[u8]> + Ord,
-	B: AsRef<[u8]>,
+	fn trie_build<DB, I, A, B>(&self, db: &mut DB, input: I) -> <Self::Hash as Hasher>::Out where
+		DB: HashDB<Self::Hash, DBValue>,
```

Comment:
why having changed the the HashDB value from usize to DBValue ?

Reply:
Was not really working with usize, except for tests. I think this function is used nowhere, could just be removed.